### PR TITLE
Upgrade to 2.13.0

### PR DESCRIPTION
**Pull request by kabentley**
update to use new ElectronHost/ElectronApp apis

### CODE REVIEW

**Comment by kabentley** [comment]
@calebmshafer i don't understand the lint errors

File: src/common/ViewerProps.ts [hunk 1]
```diff
@@ -21,10 +16,15 @@ export interface ViewerConfig {
   redirectUri: string;
 }
 
-/**
- * Returns a list of RPCs supported by this application
- */
-export function getSupportedRpcs(): RpcInterfaceDefinition[] {
+// ipc channel name for this application
+export const dtsChannel = "dts";
```
Comment (williamkbentley) [inline src/common/ViewerProps.ts:20]: dts = Desktop Starter? That is not immediately obvious to me.

Comment (kabentley) [reply src/common/ViewerProps.ts:20]: yeah, I copied this from "Display Test App" where someone used "dta". Probably should be "starterChannel" or something.

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,20 +1,91 @@
 // Place your settings in this file to overwrite default and user settings.
 {
-  "editor.detectIndentation": false,
-  "editor.insertSpaces": true,
   "editor.tabSize": 2,
-  "editor.trimAutoWhitespace": true,
+  "editor.insertSpaces": true,
```
Comment (williamkbentley) [inline .vscode/settings.json:4]: This seems like cruft.

**Comment by williamkbentley** [comment]
> @calebmshafer i don't understand the lint errors

I agree, but this one seems more direct:

  385:26  error  'RemoteBriefcaseConnection' is deprecated. use BriefcaseConnection with an IpcApp  deprecation/deprecation

File: src/common/ViewerProps.ts (hunk 1 above)
Comment (kabentley) [reply src/common/ViewerProps.ts]: fixed

File: .vscode/settings.json (hunk 2 above)
Comment (kabentley) [reply .vscode/settings.json:4]: i copied this whole file from our main repo. `insertSpaces` tells the editor to never use tabs.

Comment (williamkbentley) [reply .vscode/settings.json:4]: I was referring to all changes in this file. Doesn't seem necessary. For example, `"**/generated-docs": true,` doesn't even exist in this repo.

**Comment by kabentley** [comment]
> > @calebmshafer i don't understand the lint errors
> 
> I agree for the "rule not found" ones, but this one seems more direct:
> 
> \desktop-starter\src\frontend\components\AppComponent.tsx
> 385:26 error 'RemoteBriefcaseConnection' is deprecated. use BriefcaseConnection with an IpcApp deprecation/deprecation

yes, this app should be reworked to use either CheckpointConnection (if it is meant to only ever open checkpoints) or BriefcaseConnection if it is meant to open/edit Briefcases. For now, we can add ignore to this warning.

File: .vscode/settings.json (hunk 2 above)
Comment (kabentley) [reply .vscode/settings.json:4]: probably a good idea to leave them in anyway, in case people copy this repo as a starter for a real project.

**Comment by kabentley** [comment]
> > > @calebmshafer i don't understand the lint errors
> > 
> > 
> > I agree for the "rule not found" ones, but this one seems more direct:
> > \desktop-starter\src\frontend\components\AppComponent.tsx
> > 385:26 error 'RemoteBriefcaseConnection' is deprecated. use BriefcaseConnection with an IpcApp deprecation/deprecation
> 
> yes, this app should be reworked to use either CheckpointConnection (if it is meant to only ever open checkpoints) or BriefcaseConnection if it is meant to open/edit Briefcases. For now, we can add ignore to this warning.

is @kckst8 working on this app?

**Comment by calebmshafer** [comment]
> > > > @calebmshafer i don't understand the lint errors
> > > 
> > > 
> > > I agree for the "rule not found" ones, but this one seems more direct:
> > > \desktop-starter\src\frontend\components\AppComponent.tsx
> > > 385:26 error 'RemoteBriefcaseConnection' is deprecated. use BriefcaseConnection with an IpcApp deprecation/deprecation
> > 
> > 
> > yes, this app should be reworked to use either CheckpointConnection (if it is meant to only ever open checkpoints) or BriefcaseConnection if it is meant to open/edit Briefcases. For now, we can add ignore to this warning.
> 
> is @kckst8 working on this app?

@kabentley he is not currently. Any work done so far has been in the other repo that we have yet to make public.  Once 2.13 is released, we may be in a state where we can finally make that repo public.  However, I'll need to talk with @kckst8 since there are probably changes he needs to make to the iTwin Viewer reacting to the latest changes.

File: .vscode/settings.json (hunk 2 above)
Comment (williamkbentley) [reply .vscode/settings.json:4]: Ok. One could argue that if a user was starting a real project they may want to start with their *own* set of rules. (I personally dislike hiding `node_modules` in the files tree as it quickly lets me know if I've done `npm i` or not)

**Comment by calebmshafer** [comment]
> @calebmshafer i don't understand the lint errors

@kabentley I'll take a look.

**Review by williamkbentley — APPROVED** [review]
Approved, pending lint errors

**Comment by calebmshafer** [comment]
> > @calebmshafer i don't understand the lint errors
> 
> @kabentley I'll take a look.

@kabentley just pushed a fix for it.  Apparently you need to have both `@bentley/eslint-plugin` and `@typescript-eslint/eslint-plugin` installed directly for some of the rules to work.  Not quite sure why that's necessary, hopefully @paulius-valiunas can shed some light on it.

**Comment by kabentley** [comment]
> just pushed a fix for it. Apparently you need to have both `@bentley/eslint-plugin` and `@typescript-eslint/eslint-plugin` installed directly for some of the rules to work. Not quite sure why that's necessary, hopefully @paulius-valiunas can shed some light on it.

@calebmshafer it now says: 

Line 385:5:  Definition for rule 'deprecation/deprecation' was not found  deprecation/deprecation

any ideas?

**Comment by paulius-valiunas** [comment]
The lint error is just webpack being weird. If you run `npm run lint` separately, it won't show any errors.

For some reason webpack insists on using its own built in eslint config which doesn't contain `deprecation` or any other plugins that we use. There is a way to make it use any config you provide, but I had too many problems with that as it doesn't seem to load things correctly when a config from an external package is used. The workaround we have in the imodeljs monorepo is to provide it a blank config that only loads the plugins but no rules, just so that it doesn't throw these "rule not found" errors on every eslint-disable comment. As a pleasant side effect, it makes the build faster by not running eslint twice. I just pushed it to this branch.

**Comment by kabentley** [comment]
@calebmshafer @wgoehrig now webpack is crashing after 11 minutes on my box.

> desktop-starter@1.0.5 build:frontend D:\temp\desktop-starter
> cross-env EXTEND_ESLINT=true react-scripts build

Creating an optimized production build...
(node:9628) ExperimentalWarning: Package name self resolution is an experimental feature. This feature could change at any time

<--- Last few GCs --->

[9628:00000256F56DCA30]   158529 ms: Mark-sweep 2042.8 (2078.4) -> 2039.5 (2059.7) MB, 164.2 / 0.1 ms  (+ 568.5 ms in 119 steps since start of marking, biggest step 11.5 ms, walltime since start of marking 773 ms) (average mu = 0.103, current mu = 0.052)

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 00007FF74AFB4FAD]
    1: StubFrame [pc: 00007FF74AFB5EDD]
Security context: 0x0312205408d1 <JSObject>
    2: /* anonymous */(aka /* anonymous */) [0000019761F0C1D1] [D:\temp\desktop-starter\node_modules\webpack-sources\lib\applySourceMap.js:~58] [pc=000003744712A8EA](this=0x012b60f804b1 <undefined>,0x0099ee44f751 <String[8]: default:>,0x004b5cc10df9 <Object map = 00000248C8184179>)
    3: SourceNode_walk [000002208EF...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

Writing Node.js report to file: report.20210217.134558.9628.0.001.json
Node.js report completed
 1: 00007FF74A3B380F napi_wrap+128063
 2: 00007FF74A3529B6 public: bool __cdecl v8::base::CPU::has_sse(void)const __ptr64+35142
 3: 00007FF74A353676 public: bool __cdecl v8::base::CPU::has_sse(void)const __ptr64+38406
 4: 00007FF74AB6A11E private: void __cdecl v8::Isolate::ReportExternalAllocationLimitReached(void) __ptr64+94
 5: 00007FF74AB521F1 public: class v8::SharedArrayBuffer::Contents __cdecl v8::SharedArrayBuffer::Externalize(void) __ptr64+833
 6: 00007FF74AA1E74C public: static void __cdecl v8::internal::Heap::EphemeronKeyWriteBarrierFromCode(unsigned __int64,unsigned __int64,class v8::internal::Isolate * __ptr64)+1436
 7: 00007FF74AA299A0 public: void __cdecl v8::internal::Heap::ProtectUnprotectedMemoryChunks(void) __ptr64+1312
 8: 00007FF74AA264C4 public: static bool __cdecl v8::internal::Heap::PageFlagsAreConsistent(class v8::internal::HeapObject)+3204
 9: 00007FF74AA1BCE3 public: bool __cdecl v8::internal::Heap::CollectGarbage(enum v8::internal::AllocationSpace,enum v8::internal::GarbageCollectionReason,enum v8::GCCallbackFlags) __ptr64+1283
10: 00007FF74AA1A354 public: void __cdecl v8::internal::Heap::AddRetainedMap(class v8::internal::Handle<class v8::internal::Map>) __ptr64+2452
11: 00007FF74AA3B53D public: class v8::internal::Handle<class v8::internal::HeapObject> __cdecl v8::internal::Factory::NewFillerObject(int,bool,enum v8::internal::AllocationType,enum v8::internal::AllocationOrigin) __ptr64+61
12: 00007FF74A7A0E21 public: class v8::internal::interpreter::JumpTableTargetOffsets::iterator & __ptr64 __cdecl v8::internal::interpreter::JumpTableTargetOffsets::iterator::operator=(class v8::internal::interpreter::JumpTableTargetOffsets::iterator && __ptr64) __ptr64+1665
13: 00007FF74AFB4FAD public: virtual bool __cdecl v8::internal::SetupIsolateDelegate::SetupHeap(class v8::internal::Heap * __ptr64) __ptr64+546637
14: 00007FF74AFB5EDD public: virtual bool __cdecl v8::internal::SetupIsolateDelegate::SetupHeap(class v8::internal::Heap * __ptr64) __ptr64+550525
15: 000003744712A8EA

**Comment by calebmshafer** [comment]
@kabentley @williamkbentley this _should_ be working now.  Let me know if you run into any other issues.

I no longer get any problems when building and was able to successfully run the app.

**Comment by kabentley** [comment]
> I no longer get any problems when building and was able to successfully run the app.

But the build is still failing...

**Comment by calebmshafer** [comment]
> > I no longer get any problems when building and was able to successfully run the app.
> 
> But the build is still failing...

@kabentley let me try it out on a lower speced machine to see if I can reproduce.

**Comment by williamkbentley** [comment]
@kabentley Can you review? I implemented CheckpointConnection and made some changes to `npm start` script

**Comment by calebmshafer** [comment]
@williamkbentley for some reason I get a blank screen on initial startup.  I wonder if there's something strange going on with pulling up a previously cached iModel.  Once I refresh though it starts working again.  Possibly something that existed prior to this PR.

**Comment by kabentley** [comment]
> @kabentley Can you review? I implemented CheckpointConnection and made some changes to `npm start` script

No, this isn't right. This is what i've been trying to get @kckst8 or someone to fix. Desktop apps should acquire a briefcase and then use `BriefcaseConnection` (`CheckpointConnection` is only for web apps connected to remote backends - that's what `openRemote` means). There needs to be some way to request to synch your local briefcase to newer versions. Let me add some code that shows what i mean.

**Comment by kabentley** [comment]
@williamkbentley @calebmshafer @kckst8 

I roughed in how i think this should work. I need to work on the backend to support pullAndMerge on readonly briefcases, but you should at least be able to get this started now.

**Comment by kabentley** [comment]
@calebmshafer why does the linter here work differently than in the monorepo. Also, the build process seems to hide all errors.

File: src/frontend/components/AppComponent.tsx [hunk 3]
```diff
@@ -5,25 +5,22 @@
...
 import "./AppComponent.css";
-
 import * as React from "react";
 import { Provider } from "react-redux";
-
-import { Config, Id64, OpenMode } from "@bentley/bentleyjs-core";
+import { Config, GuidString, Id64 } from "@bentley/bentleyjs-core";
 import { IModelHubClient, VersionQuery } from "@bentley/imodelhub-client";
-import { IModelVersion } from "@bentley/imodeljs-common";
+import { BriefcaseIdValue } from "@bentley/imodeljs-backend";
```
Comment (wgoehrig) [inline src/frontend/components/AppComponent.tsx:12]: @kabentley I did a bunch of profiling, and it looks like this import is causing webpack to pull in all of imodeljs-backend - that's probably the cause of most of the webpack issues...

Should this enum be moved to imodeljs-common?

Comment (wgoehrig) [reply src/frontend/components/AppComponent.tsx]: Yeah, I have a webpack plugin to flag backend imports from frontend code (and vice versa), but we must not have added that to our create-react-app fork... I'll look into whether we can add that back in there...

I'm also working on some sass-loader changes that should significantly speedup the build time here, but that will at a minimum require us to publish a new react-scripts version.  Hopefully we can try that here by the end of the day.

Comment (wgoehrig) [reply src/frontend/components/AppComponent.tsx]: Just pushed my build performance fixes.  By disabling ESLint, Babel (for node_modules), and Terser, it looks like we've got the CI build time down from [5m 25s](https://github.com/imodeljs/desktop-starter/pull/47/checks?check_run_id=2007711063) to [2m 0s](https://github.com/imodeljs/desktop-starter/pull/47/checks?check_run_id=2032371495).

**Comment by calebmshafer** [comment]
@kabentley @williamkbentley @wgoehrig everyone mind re-reviewing?  This PR should be good to go now.

@kabentley I had to switch back to use function properties in order for the `this.setState` to work correctly. Not entirely sure why, someone with more React experience would need to fill in the details...

Either way, hoping that we can replace this whole `AppComponent` with a new functional component that uses the iTwin Viewer shortly anyway.

**Comment by williamkbentley** [comment]
> @kabentley @williamkbentley @wgoehrig everyone mind re-reviewing? This PR should be good to go now.
> 
> @kabentley I had to switch back to use function properties in order for the `this.setState` to work correctly. Not entirely sure why, someone with more React experience would need to fill in the details...
> 
> Either way, hoping that we can replace this whole `AppComponent` with a new functional component that uses the iTwin Viewer shortly anyway.

There is still a TODO in AppComponent.

**Comment by kabentley** [comment]
@calebmshafer i made a different fix, that i think is better. But, hopefully we replace the whole thing soon.

**Comment by calebmshafer** [comment]
> @calebmshafer i made a different fix, that i think is better. But, hopefully we replace the whole thing soon.

@kabentley sounds good. You okay with me reverting this PR to 2.13? I'd like to get it merged once we release 2.13 (hopefully today).

**Comment by kabentley** [comment]
> @kabentley sounds good. You okay with me reverting this PR to 2.13? I'd like to get it merged once we release 2.13 (hopefully today).

Oh sorry, I was going to try to use a change i made recently in 2.14 to eliminate SyncMode, but it's not there yet anyway. 2.13 is fine.

**Review by wgoehrig — APPROVED** [review]

**Comment by calebmshafer** [comment]
@williamkbentley mind taking a look and approving?

**Review by williamkbentley — APPROVED** [review]
LGTM. With a reminder that we need to come back to the `briefcase.pullAndMergeChanges()` TODO that was removed.

**Pull request merged**